rundetection/rules/sans_rules.py

```diff
@@ -1,173 +1,305 @@
+
+
+def _get_scatter_title(scatter: SansFileData | str | None) -> str:
+    if scatter is None:
+        return ""
+    if isinstance(scatter, SansFileData):
+        scatter = scatter.title
+    if "}_{" in scatter:
+        return scatter.split("}_{")[0].lstrip("{").rstrip("}")
+    return scatter.split("}_")[0].lstrip("{").rstrip("}")
+
+
+def _get_background_title(scatter_title: str) -> str | None:
+    if "}_{" in scatter_title:
+        title = scatter_title.split("}_{")[1]
+        # Strip off the _SANS/TRANS or _SANS from the title
+        title = title[: len(title) - 1 - len(title.split("_")[-1])]
+        return title.lstrip("{").rstrip("}")
     return None
 
 
-def _find_can_scatter_file(sans_files: list[SansFileData], can_title: str) -> SansFileData | None:
-    for sans_file in sans_files:
-        if _is_can_scatter_file(sans_file=sans_file, can_title=can_title):
-            return sans_file
+def _find_trans(job_request: JobRequest, scatter: SansFileData | None) -> SansFileData | None:
+    if job_request.additional_values["included_trans_as_scatter"]:
+        # Trans is the scatter file
+        return scatter
+    return _find_file_in_journal_by_title_and_type(title=_get_scatter_title(scatter), file_types={"TRANS"})
+
+
+def _is_direct_file(job_request: JobRequest, file_path: Path | None) -> bool:
+    if file_path is None:
+        return False
+    possible_file = load_h5py_dataset(file_path)
+    instrumentation_dict = job_request.additional_values["instrument_direct_file_comparison"]
+    # Following code can only handle a 2 deep dictionary with the one item in the top dict layer,
+    # this could be refactored to be more extensible in future but unneeded at present as it is usually just selog
+    top_layer_key = next(iter(instrumentation_dict.keys()))
+    for key in instrumentation_dict[top_layer_key]:
+        if instrumentation_dict[top_layer_key][key] != possible_file.get(top_layer_key).get(key).get("value")[0]:
+            return False
+    return True
+
+
+def _grab_file_prefix(job_file_path: Path) -> str:
+    # Reverse through the path stem until it finds the 0s then once past the 0s append to the file_prefix all
+    # characters from the stem, then finally reverse the string before returning it.
+    file_prefix = ""
+    first_zero_found = False
+    last_zero_found = False
+    for char in reversed(job_file_path.stem):
+        if last_zero_found:
+            file_prefix += char
+        elif char != "0" and first_zero_found and not last_zero_found:
+            last_zero_found = True
+            file_prefix += char
+        elif char == "0":
+            first_zero_found = True
+    # Now reverse the string and return using slicing
+    return file_prefix[::-1]
+
+
+def _get_zeros(job_file_path: Path, run_number: str) -> str:
+    num_zeros = len(job_file_path.stem) - len(run_number) - len(_grab_file_prefix(job_file_path))
+    if num_zeros > 0:
+        return num_zeros * "0"
+    return ""
```

Comment:
```suggestion
    return num_zeros * "0"
```

Reply:
That was the original code, nice, clean, readable. But if it's 0 there was an issue with the returned value, I don't recall if tests or mypy/ruff. I'll try it again because it's MUCH cleaner.